### PR TITLE
executor: fix bug when use limit in CTE

### LINKED ISSUE #25258: Unexpected result when limit used in CTE
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists t;
    create table t(a int, b int);
    insert into t values (1, 1),(2, 2),(3, 3),(1, 1),(2, 2),(3, 3);
    with recursive cte(a) as (select a from t union all select a + 1 from cte limit 10) select * from cte;

### 2. What did you expect to see? (Required)

    mysql>     with recursive cte(a) as (select a from t union all select a + 1 from cte limit 10) select * from cte;
    +------+
    | a    |
    +------+
    |    1 |
    |    2 |
    |    3 |
    |    1 |
    |    2 |
    |    3 |
    |    2 |
    |    3 |
    |    4 |
    |    2 |
    +------+
    10 rows in set (0.00 sec)
### 3. What did you see instead (Required)

    mysql>     with recursive cte(a) as (select a from t union all select a + 1 from cte limit 10) select * from cte;
    +------+
    | a    |
    +------+
    |    1 |
    |    2 |
    |    3 |
    |    1 |
    |    2 |
    |    3 |
    |    2 |
    |    3 |
    |    4 |
    |    2 |
    |    3 |
    +------+
    11 rows in set (0.01 sec)

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #25258 

Problem Summary: Got logical error when handling limit in CTE. See issue page for detailed SQL.

### What is changed and how it works?

What's Changed:  change some logic in CTEExec.nextChunkLimit().

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects: None

### Release note 

- executor: fix bug when use limit in CTE

### CODE REVIEW

File: executor/cte_test.go [hunk 1]
```diff
@@ -373,4 +373,75 @@ func (test *CTETestSuite) TestCTEWithLimit(c *check.C) {
...
+	rows = tk.MustQuery("with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 from cte1 limit 0 offset 2) select * from cte1;")
+	rows.Check(testkit.Rows())
+	rows = tk.MustQuery("with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 from cte1 limit 1 offset 2) select * from cte1;")
+	rows.Check(testkit.Rows("3"))
+	rows = tk.MustQuery("with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 from cte1 limit 2 offset 2) select * from cte1;")
+	rows.Check(testkit.Rows("3", "4"))
+	rows = tk.MustQuery("with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 from cte1 limit 3 offset 2) select * from cte1;")
+	rows.Check(testkit.Rows("3", "4", "5"))
+	rows = tk.MustQuery("with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 from cte1 limit 4 offset 2) select * from cte1;")
+	rows.Check(testkit.Rows("3", "4", "5", "6"))
```
Comment (wjhuang2016) [inline executor/cte_test.go:391]: Can we force the order?

**Review by wjhuang2016 — APPROVED** [review]

File: executor/cte_test.go (hunk 1 above)
Comment (guo-shaoge) [reply executor/cte_test.go:391]: The order should be deterministic.

**Review by xiongjiwei — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**